src/v/cloud_storage/remote_partition.cc

```diff
@@ -578,17 +578,26 @@ remote_partition::aborted_transactions(offset_range offsets) {
 }
 
 ss::future<> remote_partition::stop() {
-    _state = partition_state::stopping;
     vlog(_ctxlog.debug, "remote partition stop {} segments", _segments.size());
 
+    // Remove materialized_segment_state from the list that contains it, to
+    // avoid it getting registered for eviction and stop.
+    std::for_each(_segments.begin(), _segments.end(), [this](auto& pair) {
```

Comment:
I think we should do this after the gate close: otherwise there can still be readers existing between here and the `co_await _gate.close()` that might add things to _segments.

Reply:
I actually had it that way and then later moved it to before here https://github.com/redpanda-data/redpanda/compare/04b8185be0189ea430ecbc5a71543e3cad78ee41..0a6fcd91aa68cff31ccc8e4858e97aa8a37db940 

The reason is that while the `stop` is waiting for the gate to close, A `trim_segments` run could happen and pick up some segments for eviction and offload them. 

But now thinking about it, the only thing this will cause is that both `materialized_segments` and `remote_partition` will later try to stop that segment, but we have guards against that in the remote segment stop method, so it should be ok. 

I will change it to move after the gate close.

Reply:
fixed

Reply:
> But now thinking about it, the only thing this will cause is that both materialized_segments and remote_partition will later try to stop that segment, but we have guards against that in the remote segment stop method, so it should be ok.

I think it's even better than that: because trim_segments calls remote_partition::offload_segment, which removes the segment from remote_partition::_segments, it will not have two stop calls.